### PR TITLE
Adding logic to migrate to new javascript project experience 

### PR DESCRIPTION
**Pull request by Shenniey**
Same PR as #2456, but with squashed commits

### CODE REVIEW

File: Nodejs/MigrateToJsps/packages.config
Comment (Shenniey) [inline Nodejs/MigrateToJsps/packages.config:1]: Copying [Joaquin's comment](https://github.com/microsoft/nodejstools/pull/2456#discussion_r1175795986) from #2456 over for tracking:

Not blocking the merge, but this file shouldn't exist and it should be PackageReferences instead.

File: Nodejs/MigrateToJsps/Jsps/EsprojFileModel.cs [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+using System.Xml;
+using System.Xml.Serialization;
+
+namespace MigrateToJsps
+{
+    [XmlRoot(ElementName = "Project")]
+    public class EsprojFile
+    {
+        [XmlAttribute(AttributeName = "Sdk")]
+        public string Sdk = @"Microsoft.VisualStudio.JavaScript.Sdk/0.5.64-alpha";
```
Comment (PranavSenthilnathan) [inline Nodejs/MigrateToJsps/Jsps/EsprojFileModel.cs:14]: Latest is now 0.5.85-alpha: https://www.nuget.org/packages/Microsoft.VisualStudio.JavaScript.SDK/0.5.85-alpha

File: Nodejs/Product/Nodejs/NodejsTools.vsct [hunk 2]
```diff
@@ -334,6 +356,12 @@
       <IDSymbol name="AddNewFileGroup" value="0x1016"/>
     </GuidSymbol>
 
+    <GuidSymbol name="guidMigrateToJspsCmdSet" value="{5F602F6D-4592-448A-B321-2ECFA442BB29}">
+      <IDSymbol name="cmdidJspsProjectMigrate" value="0x206" />
+      <IDSymbol name="cmdidJspsProjectRevert" value="0x207" />
+      <IDSymbol name="cmdidJspsMigrate" value="0x208" />
```
Comment (PranavSenthilnathan) [inline Nodejs/Product/Nodejs/NodejsTools.vsct:362]: I think this is for a group rather than a command so I suggest renaming it (and update other instances of the old name in this file)
```suggestion
      <IDSymbol name="JspsMigrationGroup" value="0x208" />
```

File: Nodejs/Product/Nodejs/NodejsTools.vsct [hunk 3]
```diff
@@ -59,6 +59,10 @@
         <Parent guid="guidNodeToolsWseContextMenuCmdSet" id="idmNpm"/>
       </Group>
 
+      <Group guid="guidMigrateToJspsCmdSet" id="cmdidJspsMigrate">
+        <Parent guid="guidSHLMainMenu" id="IDM_VS_CTXT_PROJNODE"/>
```
Comment (PranavSenthilnathan) [inline Nodejs/Product/Nodejs/NodejsTools.vsct:63]: Should there be another entry for the revert command?

File: Nodejs/Product/Nodejs/Guids.cs [hunk 4]
```diff
@@ -42,6 +42,8 @@ internal static class Guids
         public static readonly Guid NodejsNpmCmdSet = new Guid(NodejsNpmCmdSetString);
         public static readonly Guid NodeToolsWorkspaceCmdSet = new Guid(NodeToolsWorkspaceCmdSetString);
 
+        public static readonly Guid MigrateToJspsCmdSet = new Guid("5F602F6D-4592-448A-B321-2ECFA442BB29");
```
Comment (PranavSenthilnathan) [inline Nodejs/Product/Nodejs/Guids.cs:45]: Should update this to `new Guid(MigrateToJspsCmdSetString)` pattern for consistency with the rest of the file.

File: Nodejs/MigrateToJsps/Ntvs/NjsprojFileReader.cs [hunk 5]
```diff
@@ -0,0 +1,105 @@
...
+        public static NjsprojFileModel ProcessNjsproj(string njsprojPath)
+        {
+            var name = "";
+            var startupFile = "";
+            var projectTypeGuidsString = "";
+            var nodejsPort = "";
+
+            using (var fileStream = File.Open(njsprojPath, FileMode.Open, FileAccess.Read))
+            {
+                XmlSerializer xmlSerializer = new XmlSerializer(typeof(Project));
```
Comment (PranavSenthilnathan) [inline Nodejs/MigrateToJsps/Ntvs/NjsprojFileReader.cs:24]: `var` for consistency (`name` above is initialized with `var` for example).

File: Nodejs/MigrateToJsps/packages.config
Comment (Shenniey) [reply Nodejs/MigrateToJsps/packages.config:1]: done!

File: Nodejs/MigrateToJsps/Ntvs/NjsprojFileReader.cs [hunk 6]
```diff
@@ -0,0 +1,105 @@
...
+
+                Project ntvsProj = (Project)xmlSerializer.Deserialize(fileStream);
+
+                if (ntvsProj == null)
+                {
+                    Console.Error.WriteLine("Something went wrong");
```
Comment (PranavSenthilnathan) [inline Nodejs/MigrateToJsps/Ntvs/NjsprojFileReader.cs:30]: Either remove or write a more descriptive error.

File: Nodejs/MigrateToJsps/Ntvs/NjsprojFileReader.cs [hunk 7]
```diff
@@ -0,0 +1,105 @@
...
+                    throw new Exception("Deserialized project is null");
+                }
+
+                // get startup file and name of project
+                foreach (var propertyGroup in ntvsProj.PropertyGroup)
+                {
+                    if (!string.IsNullOrEmpty(propertyGroup.Name))
+                    {
+                        //Console.WriteLine("Name prop is: " + propertyGroup.Name);
```
Comment (PranavSenthilnathan) [inline Nodejs/MigrateToJsps/Ntvs/NjsprojFileReader.cs:39]: Clean up comments

File: Nodejs/Product/Nodejs/Nodejs.csproj [hunk 8]
```diff
@@ -1030,11 +1041,11 @@
...
   -->
   <Target Name="_CopyForCreatePkgDef" BeforeTargets="GeneratePkgDef">
-    <Copy SourceFiles="$(OutputPath)..\ProjectWizard\Microsoft.NodejsTools.ProjectWizard.dll;$(OutputPath)..\Npm\Microsoft.NodejsTools.Npm.dll;$(OutputPath)..\PressAnyKey\Microsoft.NodejsTools.PressAnyKey.exe;" DestinationFolder="$(OutputPath)">
+    <Copy SourceFiles="$(OutputPath)..\ProjectWizard\Microsoft.NodejsTools.ProjectWizard.dll;$(OutputPath)..\Npm\Microsoft.NodejsTools.Npm.dll;$(OutputPath)..\PressAnyKey\Microsoft.NodejsTools.PressAnyKey.exe;$(OutputPath)..\MigrateToJsps\MigrateToJsps.dll;" DestinationFolder="$(OutputPath)">
       <Output TaskParameter="CopiedFiles" ItemName="_CopiedForCreatePkgdef" />
     </Copy>
   </Target>
   <Target Name="_DeleteAfterCreatePkgDef" AfterTargets="GeneratePkgDef">
-    <Delete Files="@(_CopiedForCreatePkgdef)" />
+    <!--<Delete Files="@(_CopiedForCreatePkgdef)" />-->
```
Comment (PranavSenthilnathan) [inline Nodejs/Product/Nodejs/Nodejs.csproj:1049]: Should this be uncommented now?

File: Nodejs/Product/Nodejs/Commands/RevertMigrationCommand.cs [hunk 9]
```diff
@@ -0,0 +1,79 @@
...
+            EnvDTE.Project project = MigrateToJspsUtils.GetActiveProject();
+
+            string projectFilepath = project.FullName;
+            string projectFolder = Path.GetDirectoryName(projectFilepath); 
+
+            string njsprojFile = Directory.GetFiles(projectFolder).Where(file => file.EndsWith(".njsproj")).FirstOrDefault();
+
+            if (njsprojFile == null)
+            {
+                // log error, show user message
```
Comment (PranavSenthilnathan) [inline Nodejs/Product/Nodejs/Commands/RevertMigrationCommand.cs:26]: TODO

File: Nodejs/Product/Nodejs/Commands/MigrateToJspsCommand.cs [hunk 10]
```diff
@@ -0,0 +1,102 @@
...
+                    {
+                        EnvDTE.Project activeProject = MigrateToJspsUtils.GetActiveProject();
+
+                        if (MigrateToJspsUtils.MigrationIsEnabled() && MigrateToJspsUtils.ProjectFileIsNtvs(activeProject.FullName))
+                        {
+                            cmd.Visible = cmd.Enabled = true;
+
+                            if (MigrateToJspsUtils.IsTypeScriptProject(activeProject))
+                            {
+                                cmd.Text = "Convert to New TypeScript Project Experience";
```
Comment (PranavSenthilnathan) [inline Nodejs/Product/Nodejs/Commands/MigrateToJspsCommand.cs:86]: I would suggest doing another pass through this PR (can be in another PR after this gets merged) and localizing all the messages.

File: Nodejs/MigrateToJsps/Jsps/EsprojFileWriter.cs [hunk 11]
```diff
@@ -0,0 +1,118 @@
...
+
+                return packageJson;
+            }
+
+            return null;
+        }
+
+        private static string GetSdkVersion()
+        {
+            // TODO: figure out if sdk is installed on machine?
```
Comment (PranavSenthilnathan) [inline Nodejs/MigrateToJsps/Jsps/EsprojFileWriter.cs:113]: You'll need to specify the version in the esproj, but the actual download will be handled by VS. There should be some endpoints in nuget.org that can list package versions which you may be able to make a web request for.

File: Nodejs/MigrateToJsps/MigrationLibrary.cs [hunk 12]
```diff
@@ -0,0 +1,115 @@
...
+
+namespace MigrateToJsps
+{
+    public class MigrationLibrary
+    {
+        public static string Migrate(string njsprojFile, string projectDir) 
+        {
+            if (string.IsNullOrEmpty(njsprojFile))
+            {
+                // LOG ERROR!!
```
Comment (PranavSenthilnathan) [inline Nodejs/MigrateToJsps/MigrationLibrary.cs:18]: TODO

File: Nodejs/MigrateToJsps/MigrationLibrary.cs [hunk 13]
```diff
@@ -0,0 +1,115 @@
...
+
+            try
+            {
+                fullPath = Path.GetFullPath(pathToNjsproj);
+            }
+            catch (Exception ex) when (ex is ArgumentException ||
+                                       ex is ArgumentNullException ||
+                                       ex is NotSupportedException)
+            {
+                Console.WriteLine("Please input a valid filepath.");
```
Comment (PranavSenthilnathan) [inline Nodejs/MigrateToJsps/MigrationLibrary.cs:83]: Where do these get output to? If there isn't already, there should be some logging to the output window of info and/or error messages.

**Review by PranavSenthilnathan — COMMENTED** [review]
Overall looks good, just had some comments on smaller stuff.

File: Nodejs/MigrateToJsps/Jsps/EsprojFileModel.cs (hunk 1 above)
Comment (Shenniey) [reply Nodejs/MigrateToJsps/Jsps/EsprojFileModel.cs:14]: done!

File: Nodejs/Product/Nodejs/Guids.cs (hunk 4 above)
Comment (Shenniey) [reply Nodejs/Product/Nodejs/Guids.cs:45]: done

File: Nodejs/MigrateToJsps/Ntvs/NjsprojFileReader.cs (hunk 5 above)
Comment (Shenniey) [reply Nodejs/MigrateToJsps/Ntvs/NjsprojFileReader.cs:24]: done

File: Nodejs/MigrateToJsps/Ntvs/NjsprojFileReader.cs (hunk 6 above)
Comment (Shenniey) [reply Nodejs/MigrateToJsps/Ntvs/NjsprojFileReader.cs:30]: removed!

**Comment by Shenniey** [comment]
Merging now b/c insertion has already completed (the build has been tested, no regressions found). I'm going to move comments to a new PR.

File: Nodejs/Product/Nodejs/Commands/MigrateToJspsCommand.cs (hunk 10 above)
Comment (Shenniey) [reply Nodejs/Product/Nodejs/Commands/MigrateToJspsCommand.cs:86]: done in #2471

File: Nodejs/Product/Nodejs/NodejsTools.vsct (hunk 3 above)
Comment (Shenniey) [reply Nodejs/Product/Nodejs/NodejsTools.vsct:63]: I don't think a separate Group is necessary

File: Nodejs/Product/Nodejs/NodejsTools.vsct (hunk 2 above)
Comment (Shenniey) [reply Nodejs/Product/Nodejs/NodejsTools.vsct:362]: done in #2471
